File: 1 course/labs-cc/lab7/prikol.cc
#include <iostream>
#include <vector>
#include "methods.hpp"
class C_Dictionary
{
    private:
        int m_index;
        int m_product;

    public:
        C_Dictionary(int index, int product) : m_product(product), m_index(index) {}

    int getProduct() const {return m_product;}
    int getIndex() const {return m_index;}

};
void dictionary_swap(C_Dictionary *a, C_Dictionary *b)
{
    C_Dictionary temp = *a;
    *a = *b;
    *b = temp;
}


void dictionary_selection_sort(std::vector<C_Dictionary> &dictionary, int cols)
{
    for (int i = cols-1; i > 0; i--)
    {
        int smallestIndex = i;
        for (int j = i-1; j >= 0; j--) /*-->*/ if (dictionary[smallestIndex].getProduct() > dictionary[j].getProduct()) /*-->*/ smallestIndex = j;

        dictionary_swap(&dictionary[smallestIndex], &dictionary[i]);        
    }
    
}

void sort_by_product_of_nums()
{
    int cols, rows, product = 1;
    std::cout << " enter the length of cols & rows:";
    std::cin >> rows >> cols;
    int** arr = new int*[cols];
    for (int i = 0; i < rows; i++)arr[i] = new int[rows];
    
    int** solution = new int*[cols];
    for (int i = 0; i < rows; i++)solution[i] = new int[rows];

    std::vector<C_Dictionary> dictionary;

    std::cout << "Enter the elements of matrix: " << std::endl;
    input(arr,rows, cols);

   for (int i = 0; i < cols; i++)
    {
        for (int  j = 0; j < rows; j++)product*=arr[j][i];
        
        C_Dictionary t(i,product);
        dictionary.push_back(t);
        product = 1;

    }
    dictionary_selection_sort(dictionary, cols);

    for (int i = 0; i < cols; i++)
    {
        for (int j = 0; j < rows; j++) solution[j][i] = arr[j][dictionary[i].getIndex()];
        
    }
    

    std::cout << "solution: " << std::endl;
    output(solution,rows,cols);

     
    for (int i = 0; i < rows; ++i) delete[] arr[i];
    delete[] arr;

    for (int i = 0; i < rows; ++i) delete[] solution[i];
    delete[] solution;
}

int main()
{
    sort_by_product_of_nums();
    return 0;
} 

File: 1-course/cs-labs/lab7/prikol.cs
using System;

class CDictionary
{
    public int Index { get; }
    public int Product { get; }

    public CDictionary(int index, int product)
    {
        Index = index;
        Product = product;
    }
}

class Program
{
    static void DictionarySwap(ref CDictionary a, ref CDictionary b)
    {
        var temp = a;
        a = b;
        b = temp;
    }

    static void Output(int[,] arr)
    {
        int rows = arr.GetLength(0);
        int cols = arr.GetLength(1);
        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < cols; j++)
            {
                Console.Write(arr[i, j] + " ");
            }
            Console.WriteLine();
        }
    }

static void Input(int[,] arr)
{
    int rows = arr.GetLength(0);
    int cols = arr.GetLength(1);
    for (int i = 0; i < rows; i++)
    {
        string input = Console.ReadLine();
        string[] elements = input.Split(' ');
        for (int j = 0; j < cols; j++) 
        {
            arr[i, j] = int.Parse(elements[j]);
        }
    }
}
    static void Swap(ref int a, ref int b)
    {
        int temp = a;
        a = b;
        b = temp;
    }

    static void DictionarySelectionSort(CDictionary[] dictionary, int cols)
    {
        for (int i = cols - 1; i > 0; i--)
        {
            int smallestIndex = i;
            for (int j = i - 1; j >= 0; j--)
            {
                if (dictionary[smallestIndex].Product > dictionary[j].Product)
                {
                    smallestIndex = j;
                }
            }
            DictionarySwap(ref dictionary[smallestIndex], ref dictionary[i]);
        }
    }

    static void SortByProductOfNums()
    {
        Console.WriteLine("Enter the number of rows and columns:");
        string[] dimensions = Console.ReadLine().Split();
        int rows = int.Parse(dimensions[0]);
        int cols = int.Parse(dimensions[1]);

        int[,] arr = new int[rows, cols];
        int[,] solution = new int[rows, cols];

        Console.WriteLine("Enter the elements of the matrix:");
        Input(arr);

        CDictionary[] dictionary = new CDictionary[cols];

        for (int i = 0; i < cols; i++)
        {
            int product = 1;
            for (int j = 0; j < rows; j++)
            {
                product *= arr[j, i];
            }
            dictionary[i] = new CDictionary(i, product);
        }

        DictionarySelectionSort(dictionary, cols);

        for (int i = 0; i < cols; i++)
        {
            for (int j = 0; j < rows; j++)
            {
                solution[j, i] = arr[j, dictionary[i].Index];
            }
        }

        Console.WriteLine("Solution:");
        Output(solution);
    }

    static void Main()
    {
        SortByProductOfNums();
    }
}
